clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemServerTest.java

```diff
@@ -0,0 +1,1375 @@
+        ObjectStats srcStats = new ObjectStats()
+            .pathType(PathTypeEnum.OBJECT)
+            .path("existing-dir/existing.src");
+        Path srcDir = new Path("lakefs://repo/main/existing-dir");
+        expectStatObjectNotFound("repo", "main", "existing-dir");
+        expectStatObjectNotFound("repo", "main", "existing-dir/");
+        expectListing("repo", "main",
+                      ImmutablePagination.builder().prefix("existing-dir/").build(),
+                      srcStats);
+
+        Path dst = new Path("lakefs://repo/main/existingdst.file");
+        ObjectStats dstStats = new ObjectStats()
+            .pathType(PathTypeEnum.OBJECT)
+            .path("existingdst.file");
+        expectStatObject("repo", "main", "existingdst.file", dstStats);
+
+        Assert.assertFalse(fs.rename(srcDir, dst));
+    }
+
+    /**
+     * file -> existing-directory-name: rename(src.txt, existing-dstdir) -> existing-dstdir/src.txt
+     */
+    @Test
+    public void testRename_existingFileToExistingDirName() throws ApiException, IOException {
+        Path src = new Path("lakefs://repo/main/existing-dir1/existing.src");
+        ObjectStats srcStats = new ObjectStats()
+            .pathType(PathTypeEnum.OBJECT)
+            .path("existing-dir1/existing.src");
+        expectStatObject("repo", "main", "existing-dir1/existing.src", srcStats);
+
+        ObjectStats dstStats = new ObjectStats()
+            .pathType(PathTypeEnum.OBJECT)
+            .path("existing-dir2/existing.src");
+        expectFileDoesNotExist("repo", "main", "existing-dir2");
+        expectFileDoesNotExist("repo", "main", "existing-dir2/");
+        expectListing("repo", "main",
+                      ImmutablePagination.builder().prefix("existing-dir2/").build(),
+                      dstStats);
+
+        Path dst = new Path("lakefs://repo/main/existing-dir2/");
+
+        mockServerClient.when(request()
+                              .withMethod("POST")
+                              .withPath("/repositories/repo/branches/main/objects/copy")
+                              .withQueryStringParameter("dest_path", "existing-dir2/existing.src")
+                              .withBody(json(gson.toJson(new ObjectCopyCreation()
+                                                         .srcRef("main")
+                                                         .srcPath("existing-dir1/existing.src")))))
+            .respond(response()
+                     .withStatusCode(201)
+                     // Actual new dstStats will be different... but lakeFSFS doesn't care.
+                     .withBody(json(gson.toJson(dstStats))));
+        expectGetBranch("repo", "main");
+        expectDeleteObject("repo", "main", "existing-dir1/existing.src");
+
+        // Need a directory marker at the source because it's now empty!
+        expectUploadObject("repo", "main", "existing-dir1/");
+
+        Assert.assertTrue(fs.rename(src, dst));
```

Comment:
Many tests here (and their original counterparts) do not test much. Can we test the physical file at this point, similar to what we do on testCreate, for example?

Reply:
There are issues with this test, agreed.  As they do not block 1.0, I do not want to perform them here.

Note however that using a mockserver tightens up these tests!  The "smart nulls" behaviour that we used on Mockito means that many calls will fail silently and _differently_ to how they fail on a real HTTP server.